src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs

```diff
@@ -600,35 +612,67 @@ internal override object UnMarshal(CLRStackValue dsObject, System.Type expectedC
             ICollection collection = null;
             if (dsObject.IsEnumerable)
             {
+                // collection could be T[] or List<object>.
                 collection = ToICollection(dsObject, arrayType, runtimeCore);
             }
             else
             {
                 // If dsObject is non array pointer but the expectedCLRType is IEnumerable, promote the dsObject to a collection.
                 Validity.Assert(typeof(IEnumerable).IsAssignableFrom(expectedCLRType));
+                
                 var obj = primitiveMarshaler.UnMarshal(dsObject, elementType, runtimeCore);
-                collection = new ArrayList(new object[] { obj });
+                collection = new List<object> { obj };
             }
 
             if (expectedCLRType.IsGenericType && !expectedCLRType.IsInterface)
             {
                 if (!collection.GetType().IsArray)
                 {
-                    Validity.Assert(collection is ArrayList);
-                    collection = (collection as ArrayList).ToArray(elementType);
+                    collection = (collection as List<object>).ToArray();
                 }
                 return Activator.CreateInstance(expectedCLRType, new[] { collection });
```

Comment:
@pinzart90 by your argument, wouldn't this existing code also be unsafe as this relies on there being a suitable constructor for whichever type is passed as `expectedCLRType`?

Reply:
Yes, I agree that it looks a bit unsafe.
Not sure what we could do, except try to get the constructor first and check if it exists
Or we could whitelist  the expectedType (Array, ArrayList, List etc)